src/library/rssscanner.c

```diff
@@ -0,0 +1,618 @@
+      goto cleanup;
+    }
+
+ cleanup:
+  evbuffer_free(ctx.input_body);
+  free(feedurl);
+  return xml;
+}
+
+static int
+rss_xml_parse_feed(const char **feed_title, const char **feed_author, mxml_node_t *xml)
+{
+  mxml_node_t *channel;
+  mxml_node_t *node;
+
+  channel = mxmlFindElement(xml, xml, "channel", NULL, NULL, MXML_DESCEND);
+  if (!channel)
+    {
+      DPRINTF(E_LOG, L_LIB, "Invalid RSS/xml, missing 'channel' node\n");
+      return -1;
+    }
+
+  node = mxmlFindElement(channel, channel, "title", NULL, NULL, MXML_DESCEND_FIRST);
+  if (!node)
+    {
+      DPRINTF(E_LOG, L_LIB, "Invalid RSS/xml, missing 'title' node\n");
+      return -1;
+    }
+  *feed_title = mxmlGetOpaque(node);
+
+  node = mxmlFindElement(channel, channel, "itunes:author", NULL, NULL, MXML_DESCEND_FIRST);
+  *feed_author = node ? mxmlGetOpaque(node) : NULL;
+
+  return 0;
+}
+
+static int
+rss_xml_parse_item(struct rss_item_info *ri, mxml_node_t *xml, void **saveptr)
+{
+  mxml_node_t *item;
+  mxml_node_t *node;
+  const char *s;
+
+  if (*saveptr)
+    {
+      item = (mxml_node_t *)(*saveptr);
+      while ( (item = mxmlGetNextSibling(item)) )
+	{
+	  s = mxmlGetElement(item);
+	  if (s && strcmp(s, "item") == 0)
+	    break;
+	}
+      *saveptr = item;
+    }
+  else
+    {
+      item = mxmlFindElement(xml, xml, "item", NULL, NULL, MXML_DESCEND);
+      *saveptr = item;
+    }
+
+  if (!item)
+    return -1; // No more items
+
+  memset(ri, 0, sizeof(struct rss_item_info));
+
+  node = mxmlFindElement(item, item, "title", NULL, NULL, MXML_DESCEND_FIRST);
+  ri->title = mxmlGetOpaque(node);
+
+  node = mxmlFindElement(item, item, "pubDate", NULL, NULL, MXML_DESCEND_FIRST);
+  ri->pubdate = mxmlGetOpaque(node);
+
+  node = mxmlFindElement(item, item, "link", NULL, NULL, MXML_DESCEND_FIRST);
+  ri->link = mxmlGetOpaque(node);
+
+  node = mxmlFindElement(item, item, "enclosure", NULL, NULL, MXML_DESCEND_FIRST);
+  ri->url = mxmlElementGetAttr(node, "url");
+  ri->type = mxmlElementGetAttr(node, "type");
+
+  return 0;
+}
+
+static void
+mfi_metadata_fixup(struct media_file_info *mfi, struct rss_item_info *ri, const char *feed_title, const char *feed_author, uint32_t time_added)
+{
+  struct tm tm;
+
+  // Always take the meta from media file if possible; some podcasts (Apple) can
+  // use mp4 streams which tend not to have decent tags so in those cases take
+  // info from the RSS and not the stream
+  if (!mfi->artist)
+    mfi->artist = safe_strdup(feed_author);
+  if (!mfi->album)
+    mfi->album  = safe_strdup(feed_title);
+  if (!mfi->url)
+    mfi->url    = safe_strdup(ri->link);
+  if (!mfi->genre || strcmp("(186)Podcast", mfi->genre) == 0)
+    {
+      free(mfi->genre);
+      mfi->genre  = strdup("Podcast");
+    }
+
+  // Title not valid on most mp4 (it becomes the url obj) so take from RSS feed
+  if (ri->type && strncmp("video", ri->type, 5) == 0)
+    {
+      free(mfi->title);
+      mfi->title = safe_strdup(ri->title);
+    }
+
```

Comment:
I've been adding a few more podcasts and I've noticed that there are a few podcasts that publish mp3 but the mp3 od NOT incl any meta but the RSS stream has a valid title.  This means that the title of the just becomes the url.

Should we just add this:
```
  else                                                                                   
    {                                                                                    
      // Forcible take the RSS feed info if available                                    
      if (ri->title)                                                                     
        {                                                                                
          free(mfi->title);                                                              
          mfi->title = safe_strdup(ri->title);                                           
        }                                                                                
    }
```
For example, this RSS: `http://feeds.simplecast.com/61hlOb8v` feed has items:
```    <item>
      <guid isPermaLink="false">2798bc8d-8702-4b8b-aad4-bb54b4ac7c0c</guid>
      <title>Takeover Talk &amp; A Camp Nou Debut</title>
      <description><![CDATA[<p>1. Takeover Latest - 3:10</p><p>2. Should a PL club need govt help to pay staff wages? - 12:30</p><p>3. The Aaron Hughes Interview - 20:50</p><p>4. A Debut in The Camp Nou - 29:00</p><p>5. Aaron on Sir Bobby - 34:45</p><p>6. Souness & Leaving St James - 45:20</p>
]]></description>
      <pubDate>Wed, 1 Apr 2020 04:00:24 +0000</pubDate>
      <author>podcasts@theathletic.com (Chris Waugh, Aaron Hughes, Taylor Payne, George Caulkin)</author>
      <link>https://theathletic.com/podcast/147-pod-on-the-tyne/?episode=19</link>
      <content:encoded><![CDATA[<p>1. Takeover Latest - 3:10</p><p>2. Should a PL club need govt help to pay staff wages? - 12:30</p><p>3. The Aaron Hughes Interview - 20:50</p><p>4. A Debut in The Camp Nou - 29:00</p><p>5. Aaron on Sir Bobby - 34:45</p><p>6. Souness & Leaving St James - 45:20</p>
]]></content:encoded>
      <enclosure length="60243424" type="audio/mpeg" url="https://cdn.simplecast.com/audio/e58d14/e58d1445-d2b5-42b8-9cdc-911c96562264/071f98c5-0fb7-4600-88e0-b804cd014a59/310320-pott-ep18_tc.mp3?aid=rss_feed"/>
      <itunes:title>Takeover Talk &amp; A Camp Nou Debut</itunes:title>
      <itunes:author>Chris Waugh, Aaron Hughes, Taylor Payne, George Caulkin</itunes:author>
      <itunes:duration>01:02:36</itunes:duration>
      <itunes:summary>This week, Newcastle fan-favourite Aaron Hughes joins Taylor Payne, Chris Waugh &amp; George Caulkin on the pod, with George reporting that a long-anticipated club takeover is now at an advanced stage...

Aaron reflects on a fantastic career, which began in the Camp Nou against Barcelona; Playing for Sir Bobby Robson, Graeme Souness, while the guys relive the anger they felt when he was sold to Aston Villa!</itunes:summary>
      <itunes:subtitle>This week, Newcastle fan-favourite Aaron Hughes joins Taylor Payne, Chris Waugh &amp; George Caulkin on the pod, with George reporting that a long-anticipated club takeover is now at an advanced stage...

Aaron reflects on a fantastic career, which began in the Camp Nou against Barcelona; Playing for Sir Bobby Robson, Graeme Souness, while the guys relive the anger they felt when he was sold to Aston Villa!</itunes:subtitle>
      <itunes:keywords>sir bobby, barcelona, newcastle, premier league, football nostalgia, exclusive interview, coronavirus, aaron hughes, football, souness, podcast</itunes:keywords>
      <itunes:explicit>yes</itunes:explicit>
      <itunes:episodeType>full</itunes:episodeType>
      <itunes:episode>18</itunes:episode>
    </item>
```
and the underlying mp3
```
$ ffprobe -hide_banner -i https://cdn.simplecast.com/audio/e58d14/e58d1445-d2b5-42b8-9cdc-911c96562264/071f98c5-0fb7-4600-88e0-b804cd014a59/310320-pott-ep18_tc.mp3
Input #0, mp3, from 'https://cdn.simplecast.com/audio/e58d14/e58d1445-d2b5-42b8-9cdc-911c96562264/071f98c5-0fb7-4600-88e0-b804cd014a59/310320-pott-ep18_tc.mp3':
  Metadata:
    date            : 2020-03-31 13:25
    encoder         : Lavf58.35.100
    id3v2_priv.XMP  : <?xpacket begin="\xef\xbb\xbf" id="W5M0MpCehiHzreSzNTczkc9d"?>\x0a<x:xmpmeta xmlns:x="adobe:ns:meta/" x:xmptk="Adobe XMP Core 5.6-c148 79.164050, 2019/10/01-18:03:16        ">\x0a <rdf:RDF xmlns:rdf="http://www.w3.org/1999/02/22-rdf-syntax-ns#">\x0a  <rdf
  Duration: 01:02:41.61, start: 0.025057, bitrate: 128 kb/s
    Stream #0:0: Audio: mp3, 44100 Hz, mono, fltp, 128 kb/s
```

Reply:
Ok, I have made that change. I think the suggestion you have can be simplified to just always using the title from the RSS if there is one. Same rule for video, so the previous special rule for video isn't required any more.